### PR TITLE
[IOSP-515] [CNSMR-3231] Documentation

### PR DESCRIPTION
**Pull request by AliSoftware**
This documentation update is focused on being **general for the reusable OSS repo.**

A [separate PR is pending in our main ios repo](https://github.com/babylonhealth/babylon-ios/pull/10427) to document the _day-to-day usage_ of that bot at Babylon specifically.

### README Documentation

This covers:
 - the main idea of how the bot works for people interested in using it as an OSS-available bot
 - some light pointers about implementation details

I didn't think it was worth it to get too much into verbose details about implementation outside of the `DispatchService` and `MergeService` overview, as code for the rest is pretty standard Vapor, but figured that the states+feedback diagrams could be a really helpful visual aid at least.

### Diagrams

The diagrams were made using the well established [GraphViz `.dot` format](https://graphviz.gitlab.io/).

I used the https://edotor.net/ online editor to easily live-edit and render those diagrams, as commented at the top of the `.dot` files

### Other changes

I took the occasion to update the README part about the `WallEView` client app, fix a compilation issue with the project. I also moved the `WallEView.xcodeproj` project inside `WallEView/` so that it would not conflict with `xed .` nor `vapor xcode -y` anymore when working on the main bot project.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,22 +18,95 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+In you need to maintain/improve the code, here are some high-level implementation details that might help you navigate the codebase
+
+### MergeService
+
+`MergeService` is a service class representing a single merge queue.
+
+ - It handles the logic of the state machine for the various states and transitions to process each Pull Request in its queue in order
+ - It is implemented using [ReactiveFeedback](https://github.com/babylonhealth/ReactiveFeedback)
+
+_(🚧 TODO: Add a diagram of the state machine 🚧)_
```
Comment (AliSoftware) [inline README.md]: [WIP] Working on this, will push it soon

Comment (AliSoftware) [reply README.md]: Done in 506b40b4e35e1b5f2e2fd26ea69edafac9c45203

File: WallEView/WallEView.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -7,6 +7,7 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		38291C5223E1D8C2005F572A /* CommitState.swift in Sources */ = {isa = PBXBuildFile; fileRef = 38291C5123E1D8C2005F572A /* CommitState.swift */; };
```
Comment (AliSoftware) [inline WallEView/WallEView.xcodeproj/project.pbxproj:10]: This file was missing from the xcodeproj – probably `CommitState` was extracted from `PullRequest.swift` into a dedicated file at some point in the repo's life but we forgot to update the WallEView client? – so I had to add it to make WallEView compile again

File: README.md [hunk 3]
```diff
@@ -18,22 +18,109 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+
+ - It handles the logic of the state machine for the various states and transitions to process each Pull Request in its queue in order
+ - It is implemented using [ReactiveFeedback](https://github.com/babylonhealth/ReactiveFeedback)
+
+Below are some diagrams to help you visualise the ReactiveFeedback state machine logic implemented in `MergeService`:
+
+<details>
+<summary>🔃 State Diagram</summary>
+
+![State Diagram](assets/mergebot_states.png)
```
Comment (AliSoftware) [inline README.md:97]: [`mergebot_states.png` ⤴️](https://raw.githubusercontent.com/babylonhealth/Wall-E/olivier/IOSP-515-CNSMR-3231/assets/mergebot_states.png)

File: README.md [hunk 4]
```diff
@@ -18,22 +18,109 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+
+</details>
+
+<details>
+<summary>▶️ Action/Feedbacks Diagrams</summary>
+
+![State Diagram](assets/mergebot_feedbacks.png)
```
Comment (AliSoftware) [inline README.md]: [Open `mergebot_feedbacks.png` ⤴️](https://raw.githubusercontent.com/babylonhealth/Wall-E/olivier/IOSP-515-CNSMR-3231/assets/mergebot_feedbacks.png)

File: README.md [hunk 5]
```diff
@@ -18,22 +18,109 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+<summary>🔃 State Diagram</summary>
+
+![State Diagram](assets/mergebot_states.png)
+
+</details>
+
+<details>
+<summary>▶️ Action/Feedbacks Diagrams</summary>
+
+![Action/Feedbacks Diagrams](assets/mergebot_feedbacks.png)
```
Comment (AliSoftware) [inline README.md:104]: [`mergebot_feedbacks.png` ⤴️](https://raw.githubusercontent.com/babylonhealth/Wall-E/olivier/IOSP-515-CNSMR-3231/assets/mergebot_feedbacks.png)

File: README.md [hunk 6]
```diff
@@ -18,22 +18,109 @@ Pull Requests should meet a specific set of criteria before being merged.
...
 
-![](https://media.giphy.com/media/26ybvJNaZZKpPONEc/giphy.gif)
+It works by listening to the GitHub webhook to know when labels are being added/removed and when Status Checks are updated on PRs
 
-### Client app
+### Detection of the Merge Label
+
+When a specific label – configurable via the `MERGE_LABEL` environment variable – is added to your Pull Requests, Wall-E will:
+
+ - Add that PR to a queue dedicated to the branch this PR is targetting
```
Comment (danspndl) [inline README.md]:
```suggestion
 - Add that PR to a queue dedicated to the branch this PR is targeting
```

File: README.md [hunk 7]
```diff
@@ -18,22 +18,109 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+ - Post a comment in the PR to let you know that the PR has been queued and its position in the queue
+
+(†) Wall-E handles one merge queue per target branch to still allow PRs targetting different branches to be merged in parallel.
```
Comment (danspndl) [inline README.md]:
```suggestion
(†) Wall-E handles one merge queue per target branch to allow PRs targeting different branches to be merged in parallel.
```

File: README.md [hunk 8]
```diff
@@ -18,22 +18,109 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+
+When the bot dequeues the top PR from a queue, it will start its integration, which consists of the following steps:
+
+ - Merge the target branch back into the PR to ensure it's up-to-date if not
+ - Wait for its status checks to pass
+    - you can configure if you want the bot to only merge if _all_ statuses are green or only the _required_ ones via the `REQUIRES_ALL_STATUS_CHECKS` environment variable
+ - Once the PR is deemed ready (status checks pass, green on GitHub with the minimal number of approvals reached), it will merge the PR into the target branch (squash)
+ - Then it will go to the next PR in the queue
+
+If there is a failure at some point doring the integration of the PR – e.g. one of the (required) status check fails:
```
Comment (danspndl) [inline README.md]: typo
```suggestion
If there is a failure at some point during the integration of the PR – e.g. one of the (required) status check fails:
```

File: README.md [hunk 9]
```diff
@@ -18,22 +18,109 @@ Pull Requests should meet a specific set of criteria before being merged.
...
 To install the client app:
- - build `WallEView.xcodeproj` in Xcode and copy the app from build products directory to your applications directory or download the app attached to the [latest GitHub release](https://github.com/babylonhealth/Wall-E/releases)
- - run `defaults write com.babylonhealth.WallEView Host <application address>` to set the url to the app
- - launch the app and enjoy
+
+ - Build `WallEView/WallEView.xcodeproj` in Xcode and copy the app from build products directory to your applications directory, or download the app attached to the [latest GitHub release](https://github.com/babylonhealth/Wall-E/releases)
+ - Run `defaults write com.babylonhealth.WallEView Host <application address>` to set the url to the app
+ - Launch the app and enjoy.
+
+ Once the app has been launched, a new icon should appear on your menubar (next to clock, wifi, anc similar menubar icons).
```
Comment (danspndl) [inline README.md]:
```suggestion
 Once the app has been launched, a new icon should appear in your menubar.
```

**Review by danspndl — APPROVED** [review]
Nice diagrams, they help a lot understanding all the transitions 👏

File: README.md [hunk 10]
```diff
@@ -18,22 +18,125 @@ Pull Requests should meet a specific set of criteria before being merged.
...
-👷‍♀️ **WIP, come back later** 👷‍♂️
+Wall-E is a bot written in Swift using Vapor.
 
-![](https://media.giphy.com/media/26ybvJNaZZKpPONEc/giphy.gif)
+It works by listening to the GitHub webhook to know when labels are being added/removed and when Status Checks are updated on PRs
 
-### Client app
+### Detection of the Merge Label
+
+When a specific label – configurable via the `MERGE_LABEL` environment variable – is added to your Pull Requests, Wall-E will:
```
Comment (ilyapuchka) [inline README.md]:
```suggestion
When a specific label – configurable via the `MERGE_LABEL` environment variable – is added to your Pull Request, Wall-E will:
```

File: README.md [hunk 11]
```diff
@@ -18,22 +18,125 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+
+ - Add that PR to a queue dedicated to the branch this PR is targeting
+ - Post a comment in the PR to let you know that the PR has been queued and its position in the queue
+
+(†) Wall-E handles one merge queue per target branch to allow PRs targeting different branches to be merged in parallel.
```
Comment (ilyapuchka) [inline README.md]: we should add that these queues are synchronous or serial

File: README.md [hunk 12]
```diff
@@ -18,22 +18,125 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+ - Once the PR is deemed ready (status checks pass, green on GitHub with the minimal number of approvals reached), it will merge the PR into the target branch (squash)
+ - Then it will go to the next PR in the queue
+
+If there is a failure at some point during the integration of the PR – e.g. one of the (required) status check fails:
+
+ - the bot will post a comment with the error on the PR
+ - then it will remove the merge label
+ - and go to the next PR in the queue
+
+If you remove the merge label from a PR that was in the queue, that PR gets removed from the queue.
```
Comment (ilyapuchka) [inline README.md]:
```suggestion
If you remove the merge label from a PR that was in the queue, that PR gets removed from the queue immediately and bot goes to the next PR.
```

File: README.md [hunk 13]
```diff
@@ -18,22 +18,125 @@ Pull Requests should meet a specific set of criteria before being merged.
...
+Env Var | Description
+---|---
+`GITHUB_ORGANIZATION`<br/>`GITHUB_REPOSITORY` | The GitHub organisation and repo name this bot will watch
+`GITHUB_TOKEN` | The OAuth token to use for calls to the GitHub API
+`GITHUB_WEBHOOK_SECRET` | The webhook secret to use to validate webhook payloads
+`MERGE_LABEL` | The name of the label to use to add a PR to the queue
+`TOP_PRIORITY_LABELS` | The name of the labels to use to mark as PR as top-priority – separate multiple label names by a comma
+`REQUIRES_ALL_STATUS_CHECKS` | Defines if the bot should require _all_ status checks to be green before allowing to merge a PR, or only the ones configured as _required_ in GitHub settings (the default)
+
+Some other environment variables allow further configuration of the bot, like values vor various timeouts; for the list of them all, see [`Sources/App/Extensions/EnvironmentProperties.swift`](https://github.com/babylonhealth/Wall-E/blob/master/Sources/App/Extensions/EnvironmentProperties.swift).
```
Comment (ilyapuchka) [inline README.md]:
```suggestion
Some other environment variables allow further configuration of the bot, like values for various timeouts; for the list of them all, see [`Sources/App/Extensions/EnvironmentProperties.swift`](https://github.com/babylonhealth/Wall-E/blob/master/Sources/App/Extensions/EnvironmentProperties.swift).
```

File: README.md (hunk 5 above)
Comment (ilyapuchka) [reply README.md:104]: did you use one of these JS tools?

File: assets/mergebot_states.dot [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+        subgraph reduceIntegrating {
+            integrating -> ".integrationDidChangeStatus(.done)" -> ready
+            integrating -> ".integrationDidChangeStatus(.failed)" -> integrationFailed
+            integrating -> ".integrationDidChangeStatus(.updating)" -> runningStatusChecks
+            integrating -> ".pullRequestDidChange(.exclude)" -> ready
+            integrating -> ".retryIntegration" -> integrating
+        }
+
+        subgraph reduceRunningStatusChecks {
```
Comment (ilyapuchka) [inline assets/mergebot_states.dot:59]: this is easier to read then the image =)

**Review by ilyapuchka — APPROVED** [review]

File: assets/mergebot_states.dot (hunk 14 above)
Comment (AliSoftware) [reply assets/mergebot_states.dot:59]: 😄

**Comment by AliSoftware** [comment]
> did you use one of these JS tools?

I used http://edotor.net/ (as explained in this PR's description and also at the top of the `.dot` files I added in the PR)… which is an online editor that uses GraphVizJS under the hood yeah 😉

I wrote the `.dot` file manually though (didn't look into details into the JS tools to generate those `.dot` files from intermediate ascii-art format)

File: README.md (hunk 11 above)
Comment (AliSoftware) [reply README.md]:
```suggestion
(†) Wall-E handles one serial merge queue per target branch, to allow PRs targeting different branches to be merged in parallel. Which means that each PR targeting the same branch will be merged in order one after the other, but PRs targeting different branches can be merged independently in parallel.
```

**Pull request merged**